add test-title rule (fixes #8)

jfmengels opened the pull request:
Adds test-title rule (fixes #8)

Let me know if I need to change anything ;)

sindresorhus commented on docs/rules/test-title.md:
````diff
@@ -0,0 +1,35 @@
+```js
+const test = require('ava');
+
+test((t) => {
````
```js
test(t => {
```

sindresorhus commented on docs/rules/test-title.md:
````diff
@@ -0,0 +1,35 @@
+```js
+const test = require('ava');
+
+test('my test name', (t) => {
````
same as https://github.com/sindresorhus/eslint-plugin-ava/pull/13/files#r54228577

sindresorhus commented on docs/rules/test-title.md:
```diff
@@ -0,0 +1,35 @@
+
+You can set the option in configuration like this:
+
+"test-title": [2, "always"]
```
This should be in a code block

jfmengels replied:
Isn't the case in the `eslint` project rules, but sure, added it.

jfmengels commented:
Updated! (do not bother making the same comments on my other PRs, will apply them right away)

sindresorhus commented on rules/test-title.js:
```diff
@@ -0,0 +1,32 @@
+					var hasNoTitle = args.length !== 2 || literalTypes.indexOf(args[0].type) === -1;
+					var isOverThreshold = !ifMultiple || testCount > 1;
+					if (hasNoTitle && isOverThreshold) {
+						context.report(node, 'Test should specify a title.');
```
> Test should have a title.

jfmengels replied:
Done

sindresorhus commented:
Hah, you're fast :P

Doesn't look like this support chaining? `test.skip.cb()`. The plan is to extract https://github.com/sindresorhus/eslint-plugin-ava/pull/10/files#diff-5bdffee573ffb3ac8e41b63009aec3daR53, see https://github.com/sindresorhus/eslint-plugin-ava/issues/12.

sindresorhus commented:
@twada Would you be able to help me review these pull requests? You're the local AST expert ;)

sindresorhus commented:
// @novemberborn In case you have some interest in helping out reviewing. No worries if not.

jfmengels commented:
You're right, it doesn't support `test.skip.cb`. Should I create extract `isTestFunctionCall` from #12 in this PR, or make/wait for a new PR to do this?

sindresorhus commented:
Do it in this PR as it's needed here.

sindresorhus commented on test/test-title.js:
```diff
@@ -0,0 +1,49 @@
+				errors
+			},
+			{
+				code: header + 'test((t) => { t.pass(); t.end(); });',
```
`(t)` => `t`

jfmengels commented:
Ok, will do that (a bit later though ;) ).
I will move the `ImportDeclaration`/`VariableDeclarator` parts too as they improve the "is this a test file" detection (and will update my other PRs later too then. Maybe I should have made only 1 PR for 1 rule to begin with ;)  ).

sindresorhus commented:
> Maybe I should have made only 1 PR for 1 rule to begin with ;) ).

That would indeed have made easier for you, but would also have slowed down the feedback loop considerably, so I think this a good middle-ground. The merge conflicts should be easy to resolve.

twada commented:
> @twada Would you be able to help me review these pull requests? You're the local AST expert ;)

Okay ;)

twada commented on rules/test-title.js:
```diff
@@ -0,0 +1,32 @@
+				var name = callee.type === 'MemberExpression' ? callee.object.name : callee.name;
+				if (name === 'test') {
+					testCount++;
+					var hasNoTitle = args.length !== 2 || literalTypes.indexOf(args[0].type) === -1;
```
Test title could be any Expression (`Identifier`, `BinaryExpression`, `CallExpression`...and so on).
```js
const testName = '...';

// Identifier
test(testName, t => {
});

// BinaryExpression
test('prefix_' + testName + '_suffix', t => {
});
```
So you should just count `args.length` here.

sindresorhus replied:
To complicate it further, the title could be just a function name:

```js
test(function foo() {

});
```

`foo` is the used test title in the above.

jfmengels replied:
Ooh, I didn't know that. So in this case, `foo` would be the test name, and that would be ok. But `t => {}` and `function() {}` would not.
Maybe worth making an option forbidding the use of `function foo() {}` to specify the title name?

sindresorhus replied:
> Maybe worth making an option forbidding the use of function foo() {} to specify the title name?

Why? Like, what would be the gain of doing so?

jfmengels replied:
I think that some users (me first) would like a rule enforcing a "written" rule, that describes the tested behavior in a "string" format, rather than in a function name.

```js
// BAD
test(function shouldIncrementIfTypeIsINCREMENT(t) { ... });
test(function shouldDecrementIfTypeIsDECREMENT(t) { ... });

// GOOD
test('should increment value if type is INCREMENT', function(t) { ... });
test('should decrement value if type is DECREMENT', function(t) { ... });
```

sindresorhus replied:
Yeah, it seemed nice when I added it, but it's not very useful and creates inconsistency. Maybe we should just deprecate it in AVA (?). Hmm. Maybe hold off on supporting this pattern for now, until we decide on what to do about it. I don't want it holding this up.

jfmengels replied:
Yeah, it sounds like a cool trick at first, but not sure I would really want to see this in my codebase when strings are available and more readable. Will not implement it in this PR, but will add @twada's suggestion.

Don't know about deprecating it in AVA, why not.

jfmengels commented:
Updated. I moved a few things out of `prefer-power-assert` and into `utils`, (updated the existing rules too). I added some tests, but I don't feel comfortable using made-up nodes like I just did. An option would be to use a tool like `recast` to take a string and convert it to a node. What do you guys think?

(Also, didn't add tests for `isAvaImportDeclaration` as with made-up nodes, you'd pretty much have to copy-paste the implementation)

jfmengels commented:
Maybe put this on hold until #20 gets merged? ^^'

sindresorhus commented:
#20 got merged. Sorry about the churn... But now we have a good base to build on :)

jfmengels commented:
Don't worry about it ;)
Have removed my refactoring and used @twada's changes. It makes things a bit easier, thanks for that!

sindresorhus commented on rules/test-title.js:
```diff
@@ -0,0 +1,28 @@
+	var ava = createAvaRule();
+	return ava.merge({
+		CallExpression: function (node) {
+			if (!ava.isTestFile || !util.isTestFunctionCall(node.callee)) {
```
`if (!ava.isTestFile || !ava.currentTestNode) {`

Per https://github.com/sindresorhus/eslint-plugin-ava/pull/20/files#diff-a683eee4dfe24b3d7070e7482ded2411R10

twada replied:
Better to check node equality.
```js
if (!ava.isTestFile || ava.currentTestNode !== node) {
```
Since `if (!ava.isTestFile || !ava.currentTestNode) {` matches all CallExpressions inside `test` function.

sindresorhus commented on rules/test-title.js:
```diff
@@ -0,0 +1,28 @@
+
+/* eslint quote-props: [2, "as-needed"] */
+module.exports = function (context) {
+	var ifMultiple = context.options[0] === 'if-multiple';
```
Should we maybe use a options schema to prevent users from doing mistakes in the config? http://eslint.org/docs/developer-guide/working-with-rules#options-schemas

jfmengels replied:
:+1:

jfmengels commented:
Updated!

sindresorhus commented:
Woot. Looks good :)

jfmengels commented:
Yeah! Will start creating the PRs for the other ones :)
